torch/csrc/jit/tensorexpr/cuda_half_support.h

```diff
@@ -85,12 +85,32 @@ class CudaHalfScalarRewriter : public IRMutator {
   const Expr* mutate(const Var* v) override {
     auto it = var_map.find(v);
     if (it != var_map.end()) {
-      return it->second;
+      return new Cast(kHalf, it->second);
```

Comment:
Mind explaining why we need to cast the original var back to half, after we've already gone through the work of casting it from half to float ?

Reply:
Sure, the case in the issue above is something like this:
```
A[x] = B[x];
... more stuff using B[x]
```

Which after half replacement is like:

```
A[x] = half(float(B[x]));
... more stuff using float(B[x]);
```

Then registerization - which doesn't know that Half is not allowed:

```
half B_1 = B[x];
A[x] = B_1;
... more stuff using float(B_1)
```

There are again Halfs, so we have the CudaHalfScalarRewriter:

```
float B_1 = float(B[x]);
A[x] = B_1;
... more stuff using B_1
```

And this is where it crashes, because A[x] is Half and B_1 is Float. By wrapping B_1 with Half() we cover this case where there isn't an enclosing cast. If it's not a direct write to Half, then it would already be wrapped by a Float cast from the CudaHalfChecker.

As I write this I'm wondering why it's not possible to delay Half checking until registerization which should solve this (and other) weirdnesses. I remember trying but can't remember why it didn't work. Let me have a look.

Reply:
Oh, it was because we need to insert the half type definitions before the function begins... Still maybe easier to move all the logic in to the rewriter.